### DISCUSSION TITLE
Lesson 12 Question: DecentralizedStableCoin contract Ownable

### DISCUSSION THREAD

**Discussion by SCPassion** [Cyfrin/foundry-full-course-cu#1001] [Q&A] [answered]
Hello Cyfrin community. I have a question about the update of Ownable contract in OpenZeppelin library. 
According to the new OpenZeppelin Ownable contract, we need a constructor to define the initial address.

If I am correct, the owner of the DecentralizedStableCoin contract should be the DSCEngine.sol contract.
How can I define the constructor for the Ownable contract?
```Solidity
contract DecentralizedStableCoin is ERC20Burnable, Ownable {
    error DecentralizedStableCoin__MustBeMoreThanZero();
    error DecentralizedStableCoin__BurnAmountExceedsBalance();
    error DecentralizedStableCoin__NotZeroAddress();

    // As ERC20Burnable is also ERC20. We need to use ERC20's constructor.
    constructor()
        ERC20("DecentralizedStableCoin", "DSC")
        Ownable(address(this))
    {}
}
```
Is it like this? What should I put inside the Ownable(...)?

The following is my another guess. Is the following constructor valid? Is_DSCEngineAddress able to read in Ownable constructor()?
``` Solidity
    // As ERC20Burnable is also ERC20. We need to use ERC20's constructor.
    constructor(
        address _DSCEngineAddress
    ) ERC20("DecentralizedStableCoin", "DSC") Ownable(_DSCEngineAddress) {
        transferOwnership(_DSCEngineAddress);
    }
```

Thanks for answering.

**Comment by 0xusmanf** [comment, marked as answer]
> ```solidity
>     // As ERC20Burnable is also ERC20. We need to use ERC20's constructor.
>     constructor(
>         address _DSCEngineAddress
>     ) ERC20("DecentralizedStableCoin", "DSC") Ownable(_DSCEngineAddress) {
>         transferOwnership(_DSCEngineAddress);
>     }
> ```

This wouldn't work as you would need `DecentralizedStableCoin`'s address to deploy `DSCEngine`.

The right approach should be to pass the `msg.sender` to the Ownable constructor and then transfer the ownership to `DSCEngine` after deploying both contracts.

**Reply by ikennaumeh** [reply]
So this then ??
```
constructor (address _DSCEngine) ERC20 ("DecentralizedStableCoin", "DSC") Ownable(msg.sender) {
        transferOwnership(_DSCEngine);
    }
```

**Reply by SCPassion** [reply]
@usmanfarooq91 . Thanks for your reply. I am still not sure the way to transfer ownership from msg.sender to DSCEngine after deploying both contracts. I just started lesson 12. Maybe Patrick did explain it later this lesson. Thanks for answering.

I will come back to this discussion again later.

**Comment by theirrationalone** [comment]
You should allow only specific amount of DSC (threshold) whenever you need to exchange value and that's exactly what patrick was doing.... ;`)

**Reply by ikennaumeh** [reply]
how do you mean ?

**Reply by theirrationalone** [reply]
Keep @usmanfarooq91 's words in mind and go to patrick's `DeployDSCScript` and hit `ctrl + f` on there for `transferOwnership`.... 

And then re-read my words again... Then check for `approve` along with `vm.prank` into tests files. 

You'll get the complete picture.

**Reply by SCPassion** [reply]
Thanks for answering.

**Reply by ikennaumeh** [reply]
I just started out with Chapter 12 so having gotten to the deploy script bits @theirrationalone. So my contract for DecentralizedStableCoin to avoid compile time error is :

```
// SPDX-License-Identifier: MIT

// This is considered an Exogenous, Decentralized, Anchored (pegged), Crypto Collateralized low volitility coin

// Layout of Contract:
// version
// imports
// errors
// interfaces, libraries, contracts
// Type declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// view & pure functions

pragma solidity ^0.8.18;

import {ERC20Burnable, ERC20} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import {Ownable} from  "@openzeppelin/contracts/access/Ownable.sol";

/**
 * @title DecentralizedStableCoin
 * @author Ikenna Umeh
 * Collateral: Exogenous (ETH & BTC)
 * Minting: Algorithmic
 * Relative Stability: Pegged to USD
 * @notice This is the contract meant to be governed by DSCEngine. This contract is just the ERC20 implementaion of our stablecoin system.
 */
contract DecentralizedStableCoin is ERC20Burnable, Ownable {

    error DecentralizedStableCoin__MustBeMoreThanZero();
    error DecentralizedStableCoin__BurnAmountExceedsBalance();
    error DecentralizedStableCoin__NotZeroAddress();

    constructor (address _DSCEngine) ERC20 ("DecentralizedStableCoin", "DSC") Ownable(msg.sender) {
        transferOwnership(_DSCEngine);
    }

    function burn (uint256 _amount) public override onlyOwner {
        uint256 balance = balanceOf(msg.sender);

        if(_amount <= 0){
            revert DecentralizedStableCoin__MustBeMoreThanZero();
        }

        if(balance < _amount){
            revert DecentralizedStableCoin__BurnAmountExceedsBalance();
        }
        super.burn(_amount);
    }

    function mint(address _to, uint256 _amount) external onlyOwner returns (bool) {
        if(_to == address(0)){
            revert DecentralizedStableCoin__NotZeroAddress();
        }

        if(_amount <= 0){
            revert DecentralizedStableCoin__MustBeMoreThanZero();
        }

        _mint(_to, _amount);
        return true;
    }

}
```

**Reply by theirrationalone** [reply]
You should leave `constructor` body empty

i.e.... 

```solidity
constructor (args...) baseClassConst(args... ) ... {}
// next statements.... 
```
Idk if this works, therefore i haven't executed yet... Just given an idea to triage your problem

;`)

**Reply by ikennaumeh** [reply]
Thanks. I understand it better now
